File: 2023/src/day1/day1-main.ts

import { DAY1_INPUT_FILE_NAME } from '../constants';
import { getIntuitiveTimeString } from '../util/format-util';
import { loadDayInput } from '../util/input-util';
import { runAndTime } from '../util/timer';

type DigitMatch = {
  value: string;
  index: number;
};

export async function day1Main() {
  let part1CalibrationSum: number | undefined;
  let part2CalibrationSum: number | undefined;

  const inputLines = (await loadDayInput(DAY1_INPUT_FILE_NAME))
    .map(inputLine => inputLine.trim())
    .filter(inputLine => inputLine.length > 0)
  ;
  let funTime = await runAndTime(async () => {
    part1CalibrationSum = day1p1(inputLines);
  });
  if(part1CalibrationSum === undefined) {
    throw new Error('Undefined calibrationSum result from day 1 part 1');
  }
  console.log('calibrationSum:');
  console.log(part1CalibrationSum);
  console.log(`\n[day1p1] took: ${getIntuitiveTimeString(funTime)}`);

  funTime = await runAndTime(async () => {
    part2CalibrationSum = day1p2(inputLines);
  });
  if(part2CalibrationSum === undefined) {
    throw new Error('Undefined calibrationSum result from day 1 part 2');
  }
  console.log('calibrationSum:');
  console.log(part2CalibrationSum);
  console.log(`\n[day1p2] took: ${getIntuitiveTimeString(funTime)}`);
}

export function day1p1(inputStrs: string[]): number {
  let calibrationValues: number[];
  let calibrationSum: number;

  console.log('~ Day 1 part 1 ~');

  calibrationValues = inputStrs.map(getCalibrationValue);
  calibrationSum = calibrationValues.reduce((acc, curr) => {
    return acc + curr;
  }, 0);

  return calibrationSum;
}

export function day1p2(inputStrs: string[]): number {
  let calibrationSum: number;
  let calibrationValues: number[];

  console.log('~ Day 1 part 2 ~');

  calibrationValues = inputStrs.map(getCalibrationValueP2);
  calibrationSum = calibrationValues.reduce((acc, curr) => {
    return acc + curr;
  }, 0);
  return calibrationSum;
}

function getCalibrationValueP2(inputStr: string): number {
  /*
    Approach 1:
    First replace all instances of words with the corresponding digits.
      Do this with an array of patterns calling string.replace
    Run the resulting string through the original ge†CalibrationValue function
    SCRATCH:
      This wasn't the correct solution because the following strings:
      'oneight' => '18'
      'twone' => '21'

    Approach 2:
    Find the first and last digit by search
  */
  let firstMatch: DigitMatch | undefined;
  let lastMatch: DigitMatch | undefined;
  let digitMatches: DigitMatch[] = [];
  let firstDigit: string;
  let secondDigit: string;
  let calibrationString: string;

  let allMatches = [
    /one/g,
    /two/g,
    /three/g,
    /four/g,
    /five/g,
    /six/g,
    /seven/g,
    /eight/g,
    /nine/g,
    /[0-9]/g,
  ].reduce((acc, curr) => {
    let matches = inputStr.matchAll(curr);
    return [ ...acc, ...matches ];
  }, [] as RegExpMatchArray[]);

  for(let i = 0; i < allMatches.length; ++i) {
    let currMatch: RegExpMatchArray;
    let digitMatch: DigitMatch;
    currMatch = allMatches[i];
    if(
      (currMatch.index !== undefined)
      && (typeof currMatch[0] === 'string')
    ) {
      digitMatch = {
        value: currMatch[0],
        index: currMatch.index,
      };
      digitMatches.push(digitMatch);
    }
  }

  for(let i = 0; i < digitMatches.length; ++i) {
    let currMatch: DigitMatch;
    currMatch = digitMatches[i];
    if(
      (firstMatch === undefined)
      || (currMatch.index < firstMatch.index)
    ) {
      firstMatch = currMatch;
    }
    if(
      (lastMatch === undefined)
      || (currMatch.index > lastMatch.index)
    ) {
      lastMatch = currMatch;
    }
  }

  if(
    (firstMatch === undefined)
    || (lastMatch === undefined)
  ) {
    throw new Error(`No matches found for string: ${inputStr}`);
  }

  firstDigit = getDigitFromMatchString(firstMatch.value);
  secondDigit = getDigitFromMatchString(lastMatch.value);

  calibrationString = `${firstDigit}${secondDigit}`;

  // console.log(inputStr);
  // console.log(calibrationString);

  return +calibrationString;
}

function getDigitFromMatchString(digitMatchStr: string): string {
  let digitChar: string | undefined;
  if(!isNaN(+digitMatchStr)) {
    return digitMatchStr;
  }
  [
    'one',
    'two',
    'three',
    'four',
    'five',
    'six',
    'seven',
    'eight',
    'nine',
  ].some((numStr, idx) => {
    digitChar = `${idx + 1}`;
    return numStr === digitMatchStr;
  });
  if(digitChar === undefined) {
    throw new Error(`String is not a valid digit: ${digitMatchStr}`);
  }
  return digitChar;
}

function getCalibrationValue(inputStr: string): number {
  let numStr: string;
  let firstDigit: string;
  let secondDigit: string;
  let calibrationString: string;
  numStr = inputStr.replace(/[^0-9]/gi, '');
  firstDigit = numStr[0];
  secondDigit = numStr[numStr.length - 1];
  calibrationString = `${firstDigit}${secondDigit}`;

  return +calibrationString;
}

File: 2023/src/day1/day1main.rs

pub fn day1_part1(input_lines: &Vec<String>) -> u16 {
  let calibration_sum = input_lines.iter().fold(0, |acc, curr_line| {
    acc + get_calibration_value(&curr_line)
  });
  calibration_sum
}

pub fn day1_part2(input_lines: &Vec<String>) -> u16 {
  let mut calibration_values = Vec::new();

  // for input_line in &input_lines[0..2] {
  for input_line in input_lines {
    let calibration_value = get_calibration_value_p2(input_line);
    calibration_values.push(calibration_value);
  }
  return calibration_values.iter().fold(0, |acc, curr| {
    acc + curr
  })
  // 0
}
#[derive(Debug, Clone)]
struct WordMatch<'a> {
  word: &'a str,
  start: usize,
}

fn get_calibration_value_p2(input_line: &str) -> u16 {
  let words = [
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
  ];

  let mut all_matches = Vec::new();
  for word in words {
    input_line.match_indices(word)
      .for_each(|match_index| {
        let word_match = WordMatch {
          word: match_index.1,
          start: match_index.0
        };
        all_matches.push(word_match)
      });
  }
  for (i, c) in input_line.char_indices() {
    if c.is_digit(10) {
      let c_str = &input_line[i..i + c.len_utf8()];
      all_matches.push(WordMatch {
        word: c_str,
        start: i,
      })
    }
  }
  let mut first_match_opt: Option<WordMatch> = None;
  let mut last_match_opt: Option<WordMatch> = None;
  for curr_match in all_matches {
    if first_match_opt.is_none() {
      first_match_opt = Some(curr_match.clone())
    } else if let Some(ref first_match) = first_match_opt {
      if curr_match.start < first_match.start {
        first_match_opt = Some(curr_match.clone());
      }
    }
    if last_match_opt.is_none() {
      last_match_opt = Some(curr_match);
    } else if let Some(ref last_match) = last_match_opt {
      if curr_match.start > last_match.start {
        last_match_opt = Some(curr_match)
      }
    }
  }
  let first_match = first_match_opt.unwrap();
  let last_match = last_match_opt.unwrap();
  let first_digit = get_digit_from_match_string(&first_match.word);
  let last_digit = get_digit_from_match_string(&last_match.word);
  let calibration_str = format!("{}{}", first_digit, last_digit);
  let calibration_val: u16 = calibration_str.parse().unwrap();
  return calibration_val
  // 0
}

fn get_digit_from_match_string(match_str: &str) -> &str {
  if "0123456789".contains(match_str) {
    return match_str
  }
  match match_str {
    "one" => "1",
    "two" => "2",
    "three" => "3",
    "four" => "4",
    "five" => "5",
    "six" => "6",
    "seven" => "7",
    "eight" => "8",
    "nine" => "9",
    _ => unreachable!("invalid digit string: {}", match_str),
  }
}

fn get_calibration_value(input_str: &str) -> u16 {
  let mut num_str = String::new();
  let num_chars = "0123456789";
  for c in input_str.chars() {
    if num_chars.contains(c) {
      num_str.push(c)
    }
  }
  let first_digit = num_str.chars().next().unwrap();
  let second_digit = num_str.chars().last().unwrap();
  let mut calibration_string = String::new();
  calibration_string.push(first_digit);
  calibration_string.push(second_digit);
  let calibration_num: u16 = calibration_string.parse().unwrap();
  calibration_num
}
